### ISSUE TITLE
Documentation unclear about why the Exception.Source property would throw an exception

### ISSUE THREAD

**Issue by v-kydela** [microsoft/dotnet#1073]
The documentation for [`System.Exception.Source`](https://docs.microsoft.com/en-us/dotnet/api/system.exception.source) says it throws an `ArgumentException`:

![image](https://user-images.githubusercontent.com/41968495/60831964-a26cec80-a16f-11e9-8acd-6335171ef8c1.png)

This makes no sense because `Source` is a string property and the exception says "The object must be a runtime System.Reflection object."

I can see the source code for `Exception.Source` [here](https://referencesource.microsoft.com/#mscorlib/system/exception.cs,389):

```c#
public virtual String Source {
    #if FEATURE_CORECLR
    [System.Security.SecurityCritical] // auto-generated
    #endif
    get { 
        if (_source == null)
        {
            StackTrace st = new StackTrace(this,true);
            if (st.FrameCount>0)
            {
                StackFrame sf = st.GetFrame(0);
                MethodBase method = sf.GetMethod();

                Module module = method.Module;

                RuntimeModule rtModule = module as RuntimeModule;

                if (rtModule == null)
                {
                    System.Reflection.Emit.ModuleBuilder moduleBuilder = module as System.Reflection.Emit.ModuleBuilder;
                    if (moduleBuilder != null)
                        rtModule = moduleBuilder.InternalModule;
                    else
                        throw new ArgumentException(Environment.GetResourceString("Argument_MustBeRuntimeReflectionObject"));
                }

                _source = rtModule.GetRuntimeAssembly().GetSimpleName();
            }
        }

        return _source;
    }
    #if FEATURE_CORECLR
    [System.Security.SecurityCritical] // auto-generated
    #endif
    set { _source = value; }
}
```

So the exception would be thrown as a result of some internal typecasting failures, but it's unclear what steps a developer could take to generate that exception or what steps a developer should take to avoid that exception. The description of the `ArgumentException` does not seem related to anything a developer could be doing with the string property `Source`. Perhaps the documentation could be reworded somehow.

**Comment by SusLes** [comment]
@mairaw

**Comment by sujayvsarma** [comment]
What the meaning is, is clear to me from a reading of the source code posted above. Specifically, look inside the GETter at this code block:

```
if (moduleBuilder != null)
    rtModule = moduleBuilder.InternalModule;
else
    throw new ArgumentException(Environment.GetResourceString("Argument_MustBeRuntimeReflectionObject"));
```

The last line of the block above throws an ArgumentException with that specific message.

What the message means is that the value of `Source` must be a string that points to a reflected reference of what caused the problem. If you look at the code, when `(_source == null)` it tries navigating through the `StackFrame` to get at the `RuntimeModule` that caused the problem. 

Does this resolve your doubt?

**Comment by v-kydela** [comment]
@sujayvsarma - Well, no. According to the source code, the value of `Source` doesn't have to "be a string that points to a reflected reference of what caused the problem." `Source` has a public setter so it can be any string. The only time the exception would be thrown is if you try to access `Source` when the private field `_source` is null and the getter tries to generate a source but can't do it. The message "The object must be a runtime System.Reflection object" makes no sense from the perspective of a developer trying to access a string property.

**Comment by sujayvsarma** [comment]
Your reading is completely incorrect. 

First, refer to the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.exception.source?view=netframework-4.8). The description right at the top reads:

> Gets or sets **the name of the application or the object** that causes the error.

And again at the bottom it says:

> If the Source property is not set explicitly, the runtime automatically sets it to the **name of the assembly in which the exception originated**.

This (that it needs to be a name of the application/object where the exception originated) is the only valid value for Exception.Source. While you can very well set:

```
ex.Source = "v-kydela";
```

And the code might be happy about it, it is not a good idea at all. Applications that want to handle the error, and tries to parse the source of the exception -- or worse, logging / exception alerting systems would end up with strange strings that may or may not match up to source as the code evolves over time.

To me, the documentation and the purpose of the property has been clear and we depend on that specific meaning in our application code.

**Comment by v-kydela** [comment]
@sujayvsarma - Can you provide steps that will throw the exception?

**Comment by sujayvsarma** [comment]
Nope. Under any normal circumstances that line of code cannot be hit.

And finally.... you are asking the right question. :-)



On Jul 10, 2019 21:15, Kyle Delaney <notifications@github.com> wrote:

@sujayvsarma<https://github.com/sujayvsarma> - Can you provide steps that will throw the exception?

**Comment by v-kydela** [comment]
@sujayvsarma - Can the documentation be updated to explain that the exception cannot be thrown?

**Comment by sujayvsarma** [comment]
That's a valid ask :-)

But I have nothing to do with the documentation or this code. And I don't think anyone actually watches the issues filed on this repo...

You'll need to raise an issue from the bottom of the documentation page you linked to earlier.



On Jul 10, 2019 22:09, Kyle Delaney <notifications@github.com> wrote:

@sujayvsarma<https://github.com/sujayvsarma> - Can the documentation be updated to explain that the exception cannot be thrown?

**Comment by v-kydela** [comment]
@sujayvsarma - There is no doc feedback link at the bottom of the documentation page I linked to earlier
